include/zephyr/spinlock.h

```diff
@@ -199,16 +227,43 @@ static ALWAYS_INLINE int k_spin_trylock(struct k_spinlock *l, k_spinlock_key_t *
 
 	z_spinlock_validate_pre(l);
 #ifdef CONFIG_SMP
+#ifdef CONFIG_TICKET_SPINLOCKS
+	/* Estimate the spinlock values which will mean the unlocked state */
+	atomic_val_t ticket_val = atomic_get(&l->owner);
+	atomic_t next_ticket;
+	atomic_val_t next_ticket_val;
+
+	atomic_set(&next_ticket, ticket_val);
```

Comment:
This looks really suspicious to me.  You're getting the value of "owner" and then writing it unconditionally to "next_ticket" with no synchronization.   The following atomic_inc() is clearly not atomic, because the write is unconditional.

That is: what happens if you have two threads in exactly this code racing against each other?  You can draw paths through the execution where next_ticket is incremented either once or twice.

Reply:
Hi @andyross,

Here we just estimate how the l->owner and l->tail of the free spinlock would look like for the paricular l->owner. We don't really change the spinlock state until the CAS, we keep the estimation values on stack.
The atomic operations before the atomic_cas are done only for the sake of the values consistency. We can't just increment ticket_val since it's not how we change real l->owner and l->tail in k_spin_lock and k_spin_unlock:
the estimation should follow the same steps as the real algorithm where we use atomic operations on these fields.

The contending threads indeed may have the same ticket_val and next_ticket_val, but only the fastest thread will acquire the spinlock with 

```
if (!atomic_cas(&l->tail, ticket_val, next_ticket_val)) {
		goto busy;
	}
```

The second thread will not met the CAS condition and will return -EBUSY.

Moreover, we are not entirely fair here since in case the spinlock will be acquired AFTER we "freeze" the ticket_val with atomic_get(&l->owner) and released BEFORE we come to atomic_cas we still will identify the spinlock as locked. So sometimes k_spin_trylock will fail with ticket spinlocks in condition when original spinlock will succeed. This is a side effect of having 2 atomic variables instead of 1 for ticket spinlocks.